parser_library/src/parsing/parser_impl.cpp

```diff
@@ -14,453 +14,4602 @@
+    if (error)
+        return failure;
+    return std::make_unique<semantics::data_def_operand_inline>(std::move(d), range_from(start));
+}
+template<result_t<semantics::operand_ptr> (parser2::*first)(), result_t<semantics::operand_ptr> (parser2::*rest)()>
+std::optional<semantics::op_rem> parser2::with_model(bool reparse, bool model_allowed) requires(first != nullptr)
 {
-    diagnostic_consumer_transform diags([this](diagnostic_op d) {
-        if (diagnoser_)
-            diagnoser_->add_diagnostic(std::move(d));
-    });
-    auto [_, opcode] = *proc_status;
-    using wk = id_storage::well_known;
-    if (opcode.value == wk::SETA || opcode.value == wk::ACTR || opcode.value == wk::ASPACE || opcode.value == wk::AGO
-        || opcode.value == wk::MHELP)
-        resolve_expression(expr, context::SET_t_enum::A_TYPE);
-    else if (opcode.value == wk::SETB)
-    {
-        if (!expr->is_compatible(ca_expression_compatibility::setb))
-            diags.add_diagnostic(diagnostic_op::error_CE016_logical_expression_parenthesis(expr->expr_range));
+    const auto start = cur_pos(); // capture true beginning
+    if (eof())
+        return semantics::op_rem { .line_range = empty_range(start) };
 
-        resolve_expression(expr, context::SET_t_enum::B_TYPE);
-    }
-    else if (opcode.value == wk::AIF)
-    {
-        if (!expr->is_compatible(ca_expression_compatibility::aif))
-            diags.add_diagnostic(diagnostic_op::error_CE016_logical_expression_parenthesis(expr->expr_range));
+    if (auto [error] = handle_initial_space(reparse); error)
+        return std::nullopt;
 
-        resolve_expression(expr, context::SET_t_enum::B_TYPE);
-    }
-    else if (opcode.value == wk::SETC)
+    if (eof())
+        return semantics::op_rem { .line_range = empty_range(start) };
+
+    if (model_allowed && std::find(input.next, input.last, U'&') != input.last)
     {
-        resolve_expression(expr, context::SET_t_enum::C_TYPE);
+        auto model_parser = *this;
+        if (auto [error, result] = model_parser.try_model_ops(start); error)
+            return std::nullopt;
+        else if (result)
+            return std::move(*result);
+        holder->collector.prepare_for_next_statement();
     }
-    else if (opcode.value == wk::AREAD)
+
+    std::vector<semantics::operand_ptr> operands;
+
+    bool has_error = false;
+    if (follows<U','>())
+        operands.push_back(std::make_unique<semantics::empty_operand>(cur_pos_range()));
+    else if (auto [error, op] = (this->*first)(); (has_error = error))
+        operands.push_back(std::make_unique<semantics::empty_operand>(cur_pos_range()));
+    else
+        operands.push_back(std::move(op));
+
+    if (!has_error && ((rest == nullptr && except<U' '>()) || except<U',', U' '>()))
     {
-        // aread operand is just enumeration
+        has_error = true;
+        syntax_error_or_eof();
     }
-    else
+
+    while (!has_error && follows<U','>())
```

Comment:
Consider checking if `has_error` is `true` outside of the `while` statement

Reply:
I really did not want to introduce another nesting level... and all 3 compilers are able see through it anyway.